core/src/main/java/com/linecorp/armeria/server/ServerBuilder.java

```diff
@@ -1577,9 +1594,26 @@ public VirtualHostBuilder virtualHost(String hostnamePattern) {
      * @return {@link VirtualHostBuilder} for building the virtual host
      */
     public VirtualHostBuilder virtualHost(String defaultHostname, String hostnamePattern) {
+        requireNonNull(defaultHostname, "defaultHostname");
+        requireNonNull(hostnamePattern, "hostnamePattern");
+
+        final HostAndPort hostAndPort = HostAndPort.fromString(hostnamePattern);
+
+        validateHostnamePattern(hostAndPort.getHost());
+
+        final String normalizedHostnamePattern = normalizeHostnamePattern(hostAndPort.getHost());
+        final int port = hostAndPort.getPortOrDefault(-1);
+        for (VirtualHostBuilder virtualHostBuilder : virtualHostBuilders) {
+            if (!virtualHostBuilder.defaultVirtualHost() &&
+                virtualHostBuilder.equalsDefaultHostname(defaultHostname) &&
```

Comment:
What are your thoughts on utilizing `hostnamePattern` and `port` for the equality check and removing the `defaultHostname` check? Since `hostnamePattern` is solely used to identify the virtual host, I believe we can consider two virtual hosts as identical if their `hostnamePattern` matches.
https://github.com/line/armeria/blob/6e485eb4c578271818e5631335d33f0ed96e2859/core/src/main/java/com/linecorp/armeria/server/ServiceRouteUtil.java#L52
https://github.com/line/armeria/blob/4121cf5c8079da5a22bc33e78adb6b3b6218ef07/core/src/main/java/com/linecorp/armeria/server/DefaultServerConfig.java#L300

Reply:
What do you think of raising an exception if a `defaultHostname` is registered for the same `hostnamePattern` before?
It may be controversial that `defaultHostname` is ignored silently.

Reply:
That's a good suggestion. I think we can
- set the `defaultHostname` if the previous `virtualHostBuilder` does not have the `defaultHostname`
- raise an exception if the `defaultHostname` of the previous `virtualHostBuilder` not null and It's not the same with the parameter `defaultHostname`

Reply:
A couple thoughts regarding the proposal:
- Even before this PR, a `defaultHostname` could be ignored if there were multiple virtual hosts with the same `hostnamePattern`
- The suggested behavior tries to guarantee uniqueness of (`hostName`, `defaultHostname`, `port`). However, the mutability of `defaultHostname`, `hostnamePattern`, `port` does not guarantee this.

Rather, I propose:
- All methods that can mutate `hostnamePattern` are deprecated
  - `VirtualHostBuilder#hostnamePattern(String)` is deprecated
- All methods that do not receive a single `hostnamePattern` or `port` are deprecated
  - `ServerBuilder#virtualHost(String,String)` is deprecated, since the additional `defaultHostname` makes it unclear which `VirtualHostBuilder` is returned.
  - `ServerBuilder#withVirtualHost(Customizer)` is deprecated. It is not compatible with the `VirtualHostBuilder` reuse functionality. If needed, we could add a `ServerBuilder#withVirtualHost(String, Customizer)`, `ServerBuilder#withVirtualHost(int,Customizer)` instead.

This would make it clear that only `hostnamePattern` or `port` will be used for checking reusability of `VirtualHostBuilder`.

---

In terms of implementation plan, I propose we assume that the `@Deprecated` methods do not support `VirtualHostBuilder` reusability. We can simply remove the deprecated methods in the next major release which gives a straightforward upgrade plan as well.

Regarding this PR, I propose that this method is simply deprecated and we only support `ServerBuilder#virtualHost(String)` instead.

Thoughts @line/dx ?

Reply:
That's a good idea. 👍 It's time to cleanup the API.

Reply:
That sounds great to me.

>  If needed, we could add a ServerBuilder#withVirtualHost(String, Customizer)

It would be better than deprecation without a replacement. Some forks prefer explicit code blocks over the fluent style.

Reply:
I'm sorry for late checking.

> Regarding this PR, I propose that this method is simply deprecated and we only support ServerBuilder#virtualHost(String) instead.

I understand that I need to apply changes to `ServerBuilder#virtualHost(String)` in this PR. I'll remove the changes of `ServerBuilder#virtualHost(String, String).

Reply:
> I'll remove the changes of ServerBuilder#virtualHost(String, String).

Thanks!